File: src/test/java/no/fintlabs/cache/CacheEvictionServiceTest.java
package no.fintlabs.cache;

import no.fint.model.resource.FintResource;
import no.fint.model.resource.utdanning.elev.ElevResource;
import no.fintlabs.adapter.models.event.ResourceEvictionPayload;
import no.fintlabs.cache.Cache;
import no.fintlabs.cache.CacheEvictionService;
import no.fintlabs.cache.CacheService;
import no.fintlabs.cache.packing.PackingTypes;
import no.fintlabs.consumer.config.ConsumerConfiguration;
import no.fintlabs.consumer.resource.context.ResourceContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.scheduling.TaskScheduler;

import java.time.Duration;
import java.time.Instant;
import java.util.Collections;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class CacheEvictionServiceTest {

    @Mock
    private TaskScheduler scheduler;

    @Mock
    private CacheService cacheService;

    @Mock
    private ResourceContext resourceContext;

    @Mock
    private ConsumerConfiguration configuration;

    @InjectMocks
    private CacheEvictionService cacheEvictionService;

    @Mock
    private Cache<FintResource> cache;

    private final String resourceName = "elev";

    @BeforeEach
    public void setUp() {
        when(configuration.getDomain()).thenReturn("utdanning");
        when(configuration.getPackageName()).thenReturn("elev");

        Set<String> resourceNames = Collections.singleton(resourceName);
        when(resourceContext.getResourceNames()).thenReturn(resourceNames);

        when(cacheService.getCache(anyString())).thenReturn(cache);

        doAnswer(invocation -> {
            Runnable task = invocation.getArgument(0);
            task.run();
            return null;
        }).when(scheduler).schedule(any(Runnable.class), any(Instant.class));
    }

    @Test
    public void testTriggerEvictionSchedulesTaskAndEvictsCache() {
        ResourceEvictionPayload payload = new ResourceEvictionPayload("utdanning", "elev", "elev", "fintlabs.no");
        cacheEvictionService.triggerEviction(payload);

        ArgumentCaptor<Instant> instantCaptor = ArgumentCaptor.forClass(Instant.class);
        verify(scheduler).schedule(any(Runnable.class), instantCaptor.capture());
        Instant scheduledTime = instantCaptor.getValue();

        Instant now = Instant.now();
        Instant expectedTime = now.plus(Duration.ofMinutes(10));

        long secondsDifference = Duration.between(expectedTime, scheduledTime).abs().getSeconds();
        assertTrue(secondsDifference < 1, "Scheduled time should be roughly 10 minutes in the future");

        verify(cacheService).getCache("elev");
        verify(cache).evictOldCacheObjects();
    }

}

File: src/test/java/no/fintlabs/cache/CacheEvictionServiceTest.kt
package no.fintlabs.cache

import io.micrometer.core.instrument.simple.SimpleMeterRegistry
import io.mockk.Called
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import no.fintlabs.autorelation.RelationEventService
import no.fintlabs.consumer.config.ConsumerConfiguration
import no.fintlabs.consumer.config.OrgId
import no.fintlabs.consumer.resource.context.ResourceContext
import no.novari.fint.model.resource.FintResource
import no.novari.fint.model.resource.utdanning.vurdering.ElevfravarResource
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import kotlin.test.assertTrue

class CacheEvictionServiceTest {
    private lateinit var cacheService: CacheService
    private lateinit var relationEventService: RelationEventService
    private lateinit var consumerConfiguration: ConsumerConfiguration
    private lateinit var cacheEvictionService: CacheEvictionService
    private lateinit var resourceContext: ResourceContext

    @BeforeEach
    fun setUp() {
        resourceContext = mockk(relaxed = true)
        cacheService = CacheService(resourceContext)
        relationEventService = mockk(relaxed = true)
        consumerConfiguration =
            mockk {
                every { orgId } returns OrgId.from("org-123")
            }
        cacheEvictionService =
            CacheEvictionService(
                cacheService = cacheService,
                relationEventService = relationEventService,
                consumerConfiguration = consumerConfiguration,
                meterRegistry = SimpleMeterRegistry(),
            )
    }

    @AfterEach
    fun tearDown() {
        clearAllMocks()
    }

    @Test
    fun `eviction on empty cache or with unknown resource name does not call relationEventService`() {
        val resourceName = "unknown-resource"
        cacheEvictionService.evictExpired(resourceName, Long.MAX_VALUE)

        verify { relationEventService wasNot Called }
    }

    @Test
    fun `calls removeRelations for every evicted object`() {
        val resourceName = "elevfravar"
        val key1 = "k1"
        val key2 = "k2"

        val cache = cacheService.getCache(resourceName)
        val resource1 = ElevfravarResource()
        val resource2 = ElevfravarResource()
        cache.put(key1, resource1, 1)
        cache.put(key2, resource2, 2)
        cacheEvictionService.evictExpired(resourceName, Long.MAX_VALUE)

        verify(exactly = 1) {
            relationEventService.removeRelations(resourceName, key1, resource1)
            relationEventService.removeRelations(resourceName, key2, resource2)
        }
    }

    @Test
    fun `concurrent eviction trigger for same resource is queued and reruns with latest start timestamp`() {
        val resourceName = "elevfravar"
        val firstStartTimestamp = 10L
        val secondStartTimestamp = 20L
        val mockedCacheService = mockk<CacheService>()
        val cache = mockk<FintCache<FintResource>>(relaxed = true)
        every { mockedCacheService.getCache(resourceName) } returns cache

        val service =
            CacheEvictionService(
                cacheService = mockedCacheService,
                relationEventService = relationEventService,
                consumerConfiguration = consumerConfiguration,
                meterRegistry = SimpleMeterRegistry(),
            )

        val firstRunStarted = CountDownLatch(1)
        val allowFirstRunToFinish = CountDownLatch(1)
        val evictionRuns = AtomicInteger(0)

        every { cache.evictExpired(any()) } answers {
            if (evictionRuns.incrementAndGet() == 1) {
                firstRunStarted.countDown()
                allowFirstRunToFinish.await(2, TimeUnit.SECONDS)
            }
            emptySet()
        }

        val first = CompletableFuture.runAsync { service.evictExpired(resourceName, firstStartTimestamp) }
        assertTrue(firstRunStarted.await(2, TimeUnit.SECONDS), "First eviction run did not start in time")

        val second = CompletableFuture.runAsync { service.evictExpired(resourceName, secondStartTimestamp) }

        allowFirstRunToFinish.countDown()

        first.get(2, TimeUnit.SECONDS)
        second.get(2, TimeUnit.SECONDS)

        verify(exactly = 1) { cache.evictExpired(firstStartTimestamp) }
        verify(exactly = 1) { cache.evictExpired(secondStartTimestamp) }
    }
}
